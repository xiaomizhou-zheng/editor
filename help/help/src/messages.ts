import { defineMessages } from 'react-intl-next';

export const messages = defineMessages({
  help_loading: {
    id: 'help.loading',
    defaultMessage: 'Loading',
    description: 'Loading text',
  },
  help_header: {
    id: 'help.header',
    defaultMessage: 'Help',
    description: 'Header text',
  },
  help_header_whats_new: {
    id: 'help.header_whats_new',
    defaultMessage: `What's new`,
    description: `Header text in "What's New" mode`,
  },
  help_close: {
    id: 'help.close',
    defaultMessage: 'Close',
    description: 'Close text',
  },
  help_navigation_back: {
    id: 'help.navigation.back',
    defaultMessage: 'Back',
    description: 'Back text',
  },
  help_search_placeholder: {
    id: 'help.search.placeholder',
    defaultMessage: 'Search help articles',
    description: 'Place holder text for search input',
  },
  help_article_rating_title: {
    id: 'help.article_rating.title',
    defaultMessage: 'Was this helpful?',
    description: 'Rating article title text',
  },
  help_article_rating_option_yes: {
    id: 'help.article_rating.option.yes',
    defaultMessage: 'Yes',
    description: '"Yes" text used in the article rating options',
  },
  help_article_rating_option_no: {
    id: 'help.article_rating.option.no',
    defaultMessage: 'No',
    description: '"No" text used in the article rating options',
  },
  help_article_rating_form_title: {
    id: 'help.article_rating.form.title',
    defaultMessage: 'Tell us why you gave this rating',
    description: 'Title for the article rating form',
  },
  help_article_rating_accurate: {
    id: 'help.article_rating.accurate',
    defaultMessage: `It isn't accurate`,
    description: `Text for the "isn't accurate" option of the article rating form`,
  },
  help_article_rating_clear: {
    id: 'help.article_rating.clear',
    defaultMessage: `It isn't clear`,
    description: `Text for the "isn't clear" option of the article rating form`,
  },
  help_article_rating_relevant: {
    id: 'help.article_rating.relevant',
    defaultMessage: `It isn't relevant`,
    description: `Text for the "isn't relevant" option of the article rating form`,
  },
  help_article_rating_form_submit: {
    id: 'help.article_rating.form.submit',
    defaultMessage: `Submit`,
    description: `Text for the submit button in the article rating form`,
  },
  help_article_rating_form_Success: {
    id: 'help.article_rating.form.success',
    defaultMessage: `Thanks!`,
    description: 'Success message for the article rating form',
  },
  help_article_rating_form_failed: {
    id: 'help.article_rating.form.failed',
    defaultMessage: `We couldn't submit your feedback.`,
    description: 'Error message for the article rating form',
  },
  help_article_rating_form_failed_try_again: {
    id: 'help.article_rating.form.failed.try_again',
    defaultMessage: `Try again`,
    description: `Text for the "Try again" button of the article rating form`,
  },
  help_article_rating_form_cancel: {
    id: 'help.article_rating.form.cancel',
    defaultMessage: `Cancel`,
    description: `Text for the cancel button in the article rating form`,
  },
  help_article_rating_form_contact_me: {
    id: 'help.article_rating.form.contact_me',
    defaultMessage: `Atlassian can contact me about this feedback`,
    description:
      'Text for the "Contact me" checkbox in the article rating form',
  },
  help_article_list_item_type_whats_new: {
    id: 'help.article_list_item.type.whats_new',
    defaultMessage: `WHAT'S NEW`,
    description: `"What's new" article type title`,
  },
  help_article_list_item_type_help_article: {
    id: 'help.article_list_item.type.help-article',
    defaultMessage: `HELP ARTICLE`,
    description: 'Help article article type title',
  },
  help_related_article_title: {
    id: 'help.related_article.title',
    defaultMessage: 'Related articles',
    description: 'Title of the related articles list',
  },
  help_related_article_endpoint_error_title: {
    id: 'help.related_article.endpoint-error.title',
    defaultMessage: `We're having some trouble`,
    description:
      'Title for the message displayed when the related articles fetching fails',
  },
  help_related_article_endpoint_error_description: {
    id: 'help.related_article.endpoint-error.description',
    defaultMessage: `It's taking us longer than expected to show this content. It's probably a temporary problem.`,
    description:
      'Body for the message displayed when the related articles fetching fails',
  },
  help_related_article_endpoint_error_button_label: {
    id: 'help.related_article.endpoint-error.button-label',
    defaultMessage: `Try Again`,
    description:
      'text for the retry button displayed when the related articles fetching fails',
  },
  help_search_results_search_external_site: {
    id: 'help.search_results.search_external_site',
    defaultMessage: `Can't find what you're looking for? Try again with a different term or `,
    description:
      'Text for the button displayed at the end of the search result used to open an external search site',
  },
  help_search_results_no_results: {
    id: 'help.search_results.no_results',
    defaultMessage: `We can't find anything matching your search`,
    description: `Message displayed when the search doesn't return any results (line 1)`,
  },
  help_search_results_no_results_line_two: {
    id: 'help.search_results.no_results_line_two',
    defaultMessage: `Try again with a different term or `,
    description: `Message displayed when the search doesn't return any results (line 2)`,
  },
  help_search_results_external_site_link: {
    id: 'help.search_results.external_site_link',
    defaultMessage: `search all online help articles.`,
    description: `Text for the button used to open an external search site`,
  },
  help_article_error_title: {
    id: 'help.article_error.title',
    defaultMessage: `Something's gone wrong`,
    description: `Error message title for article loading`,
  },
  help_article_error_text: {
    id: 'help.article_error.text',
    defaultMessage: `Something prevented the page from loading. We will try to reconnect automatically.`,
    description: `Error message body for article loading`,
  },
  help_article_error_button_label: {
    id: 'help.article_error.button_label',
    defaultMessage: `Try now`,
    description:
      'text for the retry button displayed when an article fetching fails',
  },
  help_search_error: {
    id: 'help.search_error',
    defaultMessage: `We are having some trouble`,
    description: `Error message for article search loading (line 1)`,
  },
  help_search_error_line_two: {
    id: 'help.search_error_line_two',
    defaultMessage: `It's taking us longer than expected to show this content. It's probably a temporary problem.`,
    description: `Error message for article search loading (line 2)`,
  },
  help_search_error_button_label: {
    id: 'help.search_error.button_label',
    defaultMessage: `Retry`,
    description: 'text for the retry button displayed when the search fails',
  },
  help_whats_new_button_label: {
    id: 'help.whats_new.button_label',
    defaultMessage: 'Find out what’s changed in {productName}',
    description: `text for the "What's New" button`,
  },
  help_whats_new_button_label_without_product_name: {
    id: 'help.whats_new.button_label',
    defaultMessage: 'Find out what’s changed',
    description: `text for the "What's New" button`,
  },
  help_whats_new_button_tooltip: {
    id: 'help.whats_new.button_tooltip',
    defaultMessage: `See our latest feature announcements, improvements, bug fixes, and more.`,
    description: `text for the "What's New" tooltip`,
  },
  help_whats_new_filter_select_option_all: {
    id: 'help.whats_new.filter_select_option_all',
    defaultMessage: `All release types`,
    description: `text for the "Release types" dropdown`,
  },
  help_whats_new_filter_select_option_new: {
    id: 'help.whats_new.filter_select_option_new',
    defaultMessage: `New`,
    description: `text for the "Release types" dropdown option "New"`,
  },
  help_whats_new_filter_select_option_improvement: {
    id: 'help.whats_new.filter_select_option_improvement',
    defaultMessage: `Improvement`,
    description: `text for the "Release types" dropdown option "Improvement"`,
  },
  help_whats_new_filter_select_option_fix: {
    id: 'help.whats_new.filter_select_option_fix',
    defaultMessage: `Fix`,
    description: `text for the "Release types" dropdown option "Fix"`,
  },
  help_whats_new_filter_select_option_removed: {
    id: 'help.whats_new.filter_select_option_removed',
    defaultMessage: `Removed`,
    description: `text for the "Release types" dropdown option "Removed"`,
  },
  help_whats_new_filter_select_option_experiment: {
    id: 'help.whats_new.filter_select_option_experiment',
    defaultMessage: `Experiment`,
    description: `text for the "Release types" dropdown option "Experiment"`,
  },
  help_whats_new_no_results: {
    id: 'help.whats.new_no_results',
    defaultMessage: `We can’t find any recent releases of that type.`,
    description: `text to display when the "What's new" API returns an empty array of articles"`,
  },
  help_whats_new_no_results_info: {
    id: 'help.whats.new_no_results_info',
    defaultMessage: `We only display three months of recent releases.`,
    description: `text to display when the "What's new" API returns an empty array of articles"`,
  },
  help_whats_new_no_results_clear_filter_button_label: {
    id: 'help.whats.new_no_results_clear_filter_button_label',
    defaultMessage: `Clear the filter`,
    description: `text to display when the "What's new" API returns an empty array of articles"`,
  },
  help_whats_new_no_results_clear_filter_info: {
    id: 'help.whats.new_no_results_clear_filter_button_label',
    defaultMessage: ` to try again.`,
    description: `text to display when the "What's new" API returns an empty array of articles"`,
  },
  help_whats_new_related_link_support: {
    id: 'help.whats_new.related_links.support',
    defaultMessage: `Support documentation`,
    description: `text for the "Support documentation" link in the related links section of "what's new" articles`,
  },
  help_whats_new_related_link_community: {
    id: 'help.whats_new.related_links.community',
    defaultMessage: `Discuss with community`,
    description: `text for the "Discuss with community" link in the related links section of "what's new" articles`,
  },
  help_show_more_button_label_more: {
    id: 'help.show_more_button.label_more',
    defaultMessage: 'Show {numberOfItemsLeft} more {itemsType}',
    description: '"Show more" text',
  },
  help_show_more_button_label_less: {
    id: 'help.show_more_button.label_less',
    defaultMessage: 'Show less',
    description: '"Show less" text',
  },
});
