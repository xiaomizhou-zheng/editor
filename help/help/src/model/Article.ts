import { BODY_FORMAT_TYPES } from '@atlaskit/help-article';
import type { AdfDoc } from '@atlaskit/help-article';
export interface Article extends ArticleItem {
  body: string | AdfDoc;
  bodyFormat?: BODY_FORMAT_TYPES;
  relatedArticles?: ArticleItem[];
}

// TODO - Check if we need to delete this
export enum ARTICLE_ITEM_TYPES {
  topicInProduct = 'topicInProduct',
}
export interface ArticleItem {
  description?: string;
  id: string;
  lastPublished: string;
  title: string;
  type: ARTICLE_ITEM_TYPES;
  routes?: {
    routeName: string;
    routeGroup: string;
  }[];
  topicId?: string;
  productName?: string;
  href?: string;
}

export interface ArticleFeedback {
  wasHelpful: boolean;
  feedbackReason: string;
  feedbackReasonText: string;
  contactMe: boolean;
}
