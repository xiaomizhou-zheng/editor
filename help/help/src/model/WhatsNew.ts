import { BODY_FORMAT_TYPES } from '@atlaskit/help-article';
import type { AdfDoc } from '@atlaskit/help-article';

export enum WHATS_NEW_ITEM_TYPES {
  NEW_FEATURE = 'Announcement',
  IMPROVEMENT = 'Improvement',
  FIX = 'Fix',
  REMOVED = 'Removed',
  EXPERIMENT = 'Experiment',
}
export interface WhatsNewArticle extends WhatsNewArticleItem {
  description: string | AdfDoc;
  bodyFormat?: BODY_FORMAT_TYPES;
}

export interface WhatsNewArticleItem {
  title?: string;
  changeTargetSchedule?: string;
  type?: WHATS_NEW_ITEM_TYPES;
  status?: string;
  featureRolloutDate?: string;
  relatedExternalLinks?: string;
  communityUrl?: string;
  href?: string;
  id: string;
}

export interface whatsNewSearchResult {
  articles: WhatsNewArticleItem[];
  nextPage: string;
  hasNextPage: boolean;
}
