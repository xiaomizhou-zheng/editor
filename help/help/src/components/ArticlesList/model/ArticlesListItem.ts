import { ArticleItem } from '../../../model/Article';
import UIAnalyticsEvent from '@atlaskit/analytics-next/UIAnalyticsEvent';

export interface ArticlesList {
  // Style. This component has two different styles (primary and secondary)
  style?: 'primary' | 'secondary';
  /* List of articles. This prop is optional */
  articles?: ArticleItem[] | null;
  /* Minimun number of articles to display. This prop is optional (Default value = 3) */
  minItemsToDisplay?: number;
  /* Minimun number of articles to display. This prop is optional (Default value = number of articles)*/
  maxItemsToDisplay?: number;
  /* Function executed when the user clicks one of the articles */
  onArticlesListItemClick?: (
    event: React.MouseEvent<HTMLElement, MouseEvent>,
    analyticsEvent: UIAnalyticsEvent,
    articleData: ArticleItem,
  ) => void;
  /* function executed when the user clicks "Show More" */
  onToggleArticlesList?: (
    event: React.MouseEvent<HTMLElement>,
    analyticsEvent: UIAnalyticsEvent,
    isCollapsed: boolean,
  ) => void;
}
