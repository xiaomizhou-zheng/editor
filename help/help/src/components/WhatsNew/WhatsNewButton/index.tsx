import React, { useCallback } from 'react';
import { UIAnalyticsEvent, AnalyticsContext } from '@atlaskit/analytics-next';
import * as colors from '@atlaskit/theme/colors';
import LightbulbIcon from '@atlaskit/icon/glyph/lightbulb';
import { token } from '@atlaskit/tokens';
import { injectIntl, WrappedComponentProps } from 'react-intl-next';
import {
  name as packageName,
  version as packageVersion,
} from '../../../version.json';

import { useWhatsNewArticleContext } from '../../contexts/whatsNewArticleContext';
import { useNavigationContext } from '../../contexts/navigationContext';
import { messages } from '../../../messages';
import HelpContentButton from '../../HelpContentButton';
import { ARTICLE_TYPE } from '../../../model/Help';

const ANALYTICS_CONTEXT_DATA = {
  componentName: 'WhatsNewButton',
  packageName,
  packageVersion,
};

interface WhatsNewButtonProps {
  productName?: string;
}

export const WhatsNewButton: React.FC<
  WrappedComponentProps & WhatsNewButtonProps
> = ({ productName, intl: { formatMessage } }) => {
  const {
    onWhatsNewButtonClick,
    whatsNewGetNotificationProvider,
  } = useWhatsNewArticleContext();
  const { openArticle } = useNavigationContext();

  const handleOnButtonClick = useCallback(
    (
      id: string,
      analytics: UIAnalyticsEvent,
      event: React.MouseEvent<HTMLElement, MouseEvent>,
    ): void => {
      if (onWhatsNewButtonClick) {
        onWhatsNewButtonClick(event, analytics);
      }

      openArticle({ id: '', type: ARTICLE_TYPE.WHATS_NEW });
    },
    [onWhatsNewButtonClick, openArticle],
  );

  return (
    <HelpContentButton
      id="whats-new"
      key="whats-new"
      tooltipText={formatMessage(messages.help_whats_new_button_tooltip)}
      notificationLogProvider={whatsNewGetNotificationProvider}
      onClick={handleOnButtonClick}
      text={
        productName
          ? formatMessage(messages.help_whats_new_button_label, {
              productName,
            })
          : formatMessage(
              messages.help_whats_new_button_label_without_product_name,
            )
      }
      icon={
        <LightbulbIcon
          primaryColor={token('color.icon.subtle', colors.N600)}
          size="medium"
          label=""
        />
      }
    />
  );
};

const WhatsNewButtonWithContext: React.FC<
  WrappedComponentProps & WhatsNewButtonProps
> = (props) => {
  return (
    <AnalyticsContext data={ANALYTICS_CONTEXT_DATA}>
      <WhatsNewButton {...props} />
    </AnalyticsContext>
  );
};

export default injectIntl(WhatsNewButtonWithContext);
