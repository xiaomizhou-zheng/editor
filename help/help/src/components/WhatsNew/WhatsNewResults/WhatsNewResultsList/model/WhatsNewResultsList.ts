import { WhatsNewArticleItem } from '../../../../../model/WhatsNew';
import UIAnalyticsEvent from '@atlaskit/analytics-next/UIAnalyticsEvent';

export interface WhatsNewResultsList {
  // Style. This component has two different styles (primary and secondary)
  style?: 'primary' | 'secondary';
  /* List of "What's New" items. This prop is optional */
  whatsNewArticles?: WhatsNewArticleItem[] | null;
  nextPage?: string;
  hasNextPage?: boolean;
  loadingMore?: boolean;
  onWhatsNewResultItemClick?: (
    event: React.MouseEvent<HTMLElement, MouseEvent>,
    analyticsEvent: UIAnalyticsEvent,
    articleData: WhatsNewArticleItem,
  ) => void;
  onShowMoreButtonClick?: (
    event: React.MouseEvent<HTMLElement>,
    analyticsEvent: UIAnalyticsEvent,
  ) => void;
}
