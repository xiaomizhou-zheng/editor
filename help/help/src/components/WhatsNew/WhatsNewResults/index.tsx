import React, { useCallback, useState, useRef } from 'react';
import { Transition } from 'react-transition-group';
import isEqual from 'lodash/isEqual';
import Select from '@atlaskit/select';
import { injectIntl, WrappedComponentProps } from 'react-intl-next';

import { VIEW } from '../../constants';

import { messages } from '../../../messages';
import { WHATS_NEW_ITEM_TYPES } from '../../../model/WhatsNew';
import { REQUEST_STATE } from '../../../model/Requests';

import {
  FADEIN_OVERLAY_TRANSITION_DURATION_MS,
  TRANSITION_STATUS,
  NUMBER_OF_WHATS_NEW_ITEMS_PER_PAGE,
} from '../../constants';
import { useWhatsNewArticleContext } from '../../contexts/whatsNewArticleContext';
import { useNavigationContext } from '../../contexts/navigationContext';

import WhatsNewResultsEmpty from './WhatsNewResultsEmpty';
import WhatsNewResultsError from './WhatsNewResultsError';
import WhatsNewResultsLoading from './WhatsNewResultsLoading';
import WhatsNewResultsList from './WhatsNewResultsList';
import {
  WhatsNewResultsContainer,
  SelectContainer,
  WhatsNewResultsListContainer,
} from './styled';

interface SelectOption {
  value: WHATS_NEW_ITEM_TYPES | undefined | '';
  label: string;
}

const defaultStyle: Partial<{ [index: string]: string | number | null }> = {
  transition: `opacity ${FADEIN_OVERLAY_TRANSITION_DURATION_MS}ms`,
  opacity: 0,
  visibility: 'hidden',
};

const transitionStyles: {
  [id: string]: { [index: string]: string | number | null };
} = {
  entering: { opacity: 1, visibility: 'visible' },
  entered: { opacity: 1, visibility: 'visible' },
  exiting: { opacity: 0, visibility: 'visible' },
  exited: { opacity: 0, visibility: 'hidden' },
};

export const WhatsNewResults: React.FC<WrappedComponentProps> = ({
  intl: { formatMessage },
}) => {
  const { view: helpContextView } = useNavigationContext();
  const {
    searchWhatsNewArticlesResult,
    onSearchWhatsNewArticles,
    searchWhatsNewArticlesState,
    onWhatsNewResultItemClick,
  } = useWhatsNewArticleContext();
  const containerRef = useRef<HTMLDivElement>(null);
  const containerScrollPosition = useRef<number>(0);

  const SELECT_DEFAULT_VALUE: SelectOption = {
    value: '',
    label: formatMessage(messages.help_whats_new_filter_select_option_all),
  };

  const SELECT_EMPTY_VALUE: SelectOption = {
    value: undefined,
    label: '',
  };

  const [selectedOption, setSelectedOption] = useState<SelectOption>(
    SELECT_EMPTY_VALUE,
  );

  const handleOnShowMoreButtonClick = useCallback(() => {
    if (searchWhatsNewArticlesResult && onSearchWhatsNewArticles) {
      const { nextPage, hasNextPage } = searchWhatsNewArticlesResult;
      if (nextPage && hasNextPage) {
        onSearchWhatsNewArticles(
          selectedOption.value,
          NUMBER_OF_WHATS_NEW_ITEMS_PER_PAGE,
          nextPage,
        );
      }
    }
  }, [onSearchWhatsNewArticles, searchWhatsNewArticlesResult, selectedOption]);

  const handleOnEnter = () => {
    setTimeout(() => {
      if (containerRef.current) {
        containerRef.current.scrollTop = containerScrollPosition.current;
      }
    }, 0);

    if (isEqual(selectedOption, SELECT_EMPTY_VALUE)) {
      setSelectedOption(SELECT_DEFAULT_VALUE);
      onSearchWhatsNewArticles && onSearchWhatsNewArticles();
    }
  };

  const handleOnExit = () => {
    if (helpContextView === VIEW.DEFAULT_CONTENT) {
      setSelectedOption(SELECT_EMPTY_VALUE);
      containerScrollPosition.current = 0;
    } else {
      containerScrollPosition.current = containerRef.current
        ? containerRef.current.scrollTop
        : 0;
    }
  };

  const handleOnClearFilter = () => {
    if (onSearchWhatsNewArticles) {
      setSelectedOption({
        value: '',
        label: formatMessage(messages.help_whats_new_filter_select_option_all),
      });
      onSearchWhatsNewArticles('', NUMBER_OF_WHATS_NEW_ITEMS_PER_PAGE, '');
    }
  };

  return (
    <Transition
      in={helpContextView === VIEW.WHATS_NEW}
      timeout={FADEIN_OVERLAY_TRANSITION_DURATION_MS}
      onEnter={handleOnEnter}
      onExit={handleOnExit}
    >
      {(state: TRANSITION_STATUS) => (
        <WhatsNewResultsContainer
          ref={containerRef}
          style={{
            ...defaultStyle,
            ...transitionStyles[state],
          }}
        >
          <>
            {searchWhatsNewArticlesState === REQUEST_STATE.loading &&
              searchWhatsNewArticlesResult === null &&
              state !== 'exited' && <WhatsNewResultsLoading />}
            {(searchWhatsNewArticlesState === REQUEST_STATE.done ||
              searchWhatsNewArticlesState === REQUEST_STATE.loading) &&
              searchWhatsNewArticlesResult !== null &&
              state !== 'exited' && (
                <>
                  <SelectContainer>
                    <Select
                      defaultValue={selectedOption}
                      className="single-select"
                      classNamePrefix="react-select"
                      options={[
                        {
                          value: '',
                          label: formatMessage(
                            messages.help_whats_new_filter_select_option_all,
                          ),
                        },
                        {
                          value: WHATS_NEW_ITEM_TYPES.NEW_FEATURE,
                          label: formatMessage(
                            messages.help_whats_new_filter_select_option_new,
                          ),
                        },
                        {
                          label: formatMessage(
                            messages.help_whats_new_filter_select_option_improvement,
                          ),
                          value: WHATS_NEW_ITEM_TYPES.IMPROVEMENT,
                        },
                        {
                          label: formatMessage(
                            messages.help_whats_new_filter_select_option_fix,
                          ),
                          value: WHATS_NEW_ITEM_TYPES.FIX,
                        },
                        {
                          label: formatMessage(
                            messages.help_whats_new_filter_select_option_removed,
                          ),
                          value: WHATS_NEW_ITEM_TYPES.REMOVED,
                        },
                        {
                          label: formatMessage(
                            messages.help_whats_new_filter_select_option_experiment,
                          ),
                          value: WHATS_NEW_ITEM_TYPES.EXPERIMENT,
                        },
                      ]}
                      value={selectedOption}
                      onChange={(option) => {
                        if (onSearchWhatsNewArticles) {
                          const selectedOptionValue = (option as {
                            value: string;
                          }).value as WHATS_NEW_ITEM_TYPES | '';
                          const selectedOptionLabel = (option as {
                            label: string;
                          }).label as string;
                          setSelectedOption({
                            value: selectedOptionValue,
                            label: selectedOptionLabel,
                          });
                          onSearchWhatsNewArticles(
                            selectedOptionValue,
                            NUMBER_OF_WHATS_NEW_ITEMS_PER_PAGE,
                            '',
                          );
                        }
                      }}
                    />
                  </SelectContainer>
                  <WhatsNewResultsListContainer>
                    {searchWhatsNewArticlesResult.articles.length > 0 ? (
                      <WhatsNewResultsList
                        whatsNewArticles={searchWhatsNewArticlesResult.articles}
                        onWhatsNewResultItemClick={onWhatsNewResultItemClick}
                        onShowMoreButtonClick={handleOnShowMoreButtonClick}
                        hasNextPage={searchWhatsNewArticlesResult?.hasNextPage}
                        nextPage={searchWhatsNewArticlesResult?.nextPage}
                        loadingMore={
                          searchWhatsNewArticlesState === REQUEST_STATE.loading
                        }
                      />
                    ) : (
                      <WhatsNewResultsEmpty
                        onClearFilter={handleOnClearFilter}
                      />
                    )}
                  </WhatsNewResultsListContainer>
                </>
              )}

            {searchWhatsNewArticlesState === REQUEST_STATE.error && (
              <WhatsNewResultsError onSearch={onSearchWhatsNewArticles} />
            )}
          </>
        </WhatsNewResultsContainer>
      )}
    </Transition>
  );
};

export default injectIntl(WhatsNewResults);
