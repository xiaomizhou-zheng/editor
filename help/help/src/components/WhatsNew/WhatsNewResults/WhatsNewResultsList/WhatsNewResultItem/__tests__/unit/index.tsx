import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render, fireEvent } from '@testing-library/react';
import AnalyticsListener from '@atlaskit/analytics-next/AnalyticsListener';
import { createIntl, createIntlCache } from 'react-intl-next';

import { getMockWhatsNewArticleItem } from '../../../../../../../util/testing/mock';

import { WhatsNewResultListItem } from '../../index';
import { messages } from '../../../../../../../messages';
import { WHATS_NEW_ITEM_TYPES } from '../../../../../../../model/WhatsNew';

// Messages
const cache = createIntlCache();
const intl = createIntl(
  {
    locale: 'en',
    messages: {},
  },
  cache,
);
const messageTypeWhatsNewNewFeature = intl.formatMessage(
  messages.help_whats_new_filter_select_option_new,
);
const messageTypeWhatsNewFix = intl.formatMessage(
  messages.help_whats_new_filter_select_option_fix,
);

const mockOnClick = jest.fn();
const analyticsSpy = jest.fn();

describe('WhatsNewResultListItem', () => {
  it.skip('Should match snapshot', () => {
    const mockWhatsNewArticleItem = getMockWhatsNewArticleItem('1');
    const { container } = render(
      <WhatsNewResultListItem
        intl={intl}
        {...mockWhatsNewArticleItem}
        onClick={mockOnClick}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it(`Should display the type "New Feature" if the type isn't defined `, () => {
    const mockWhatsNewArticleItem = getMockWhatsNewArticleItem('1');
    const { queryByText } = render(
      <WhatsNewResultListItem
        intl={intl}
        {...mockWhatsNewArticleItem}
        onClick={mockOnClick}
      />,
    );

    const TypeHelpArticleElm = queryByText(messageTypeWhatsNewNewFeature);

    expect(TypeHelpArticleElm).not.toBeNull();
  });

  it(`Should display the correct type text if the type is defined `, () => {
    const mockWhatsNewArticleItem = getMockWhatsNewArticleItem(
      '1',
      WHATS_NEW_ITEM_TYPES.FIX,
    );
    const { queryByText } = render(
      <WhatsNewResultListItem
        intl={intl}
        {...mockWhatsNewArticleItem}
        onClick={mockOnClick}
      />,
    );

    const TypeHelpArticleElm = queryByText(messageTypeWhatsNewFix);

    expect(TypeHelpArticleElm).not.toBeNull();
  });

  it(`mockOnClick should be executed when the user clicks in the component`, () => {
    const mockWhatsNewArticleItem = getMockWhatsNewArticleItem('1');
    const { queryByRole } = render(
      <AnalyticsListener channel="help" onEvent={analyticsSpy}>
        <WhatsNewResultListItem
          intl={intl}
          {...mockWhatsNewArticleItem}
          onClick={mockOnClick}
        />
        ,
      </AnalyticsListener>,
    );

    const button = queryByRole('button');

    expect(button).not.toBeNull();
    if (button) {
      expect(mockOnClick).toHaveBeenCalledTimes(0);
      fireEvent.click(button);
      expect(mockOnClick).toHaveBeenCalledTimes(1);
    }
  });
});
