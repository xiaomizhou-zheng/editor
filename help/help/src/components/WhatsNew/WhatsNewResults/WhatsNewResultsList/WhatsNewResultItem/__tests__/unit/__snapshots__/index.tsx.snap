// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhatsNewResultListItem Should match snapshot 1`] = `
.emotion-4 {
  position: relative;
  box-sizing: border-box;
  padding: 8px;
  display: block;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  color: #6B778C;
  border-radius: 3px;
}

.emotion-4:hover,
.emotion-4:focus,
.emotion-4:visited,
.emotion-4:active {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  outline-offset: none;
}

.emotion-4:focus {
  box-shadow: #4C9AFF 0px 0px 0px 2px inset;
}

.emotion-4:hover {
  background-color: #EBECF0;
}

.emotion-4:active {
  background-color: #DEEBFF;
}

.emotion-2 {
  width: 100%;
  white-space: nowrap;
  margin-bottom: 4px;
}

.emotion-0 {
  display: inline-block;
  vertical-align: middle;
  position: relative;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  background-color: #36B37E;
}

.emotion-0 > img {
  width: 16px;
  height: 16px;
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.emotion-0 > img > svg {
  height: 100%;
  width: 100%;
}

.emotion-1 {
  font-size: 1em;
  font-style: inherit;
  line-height: 1.1428571428571428;
  color: var(--text-highEmphasis,#172B4D);
  font-weight: 600;
  -webkit-letter-spacing: -0.003em;
  -moz-letter-spacing: -0.003em;
  -ms-letter-spacing: -0.003em;
  letter-spacing: -0.003em;
  margin-top: 16px;
  display: inline-block;
  vertical-align: middle;
  margin: 0;
  padding-left: 4px;
  white-space: normal;
  overflow-x: hidden;
}

.emotion-3 {
  display: block;
  line-height: 20px;
  color: #505F79;
  margin: 0;
}

<a
  aria-disabled="false"
  class="emotion-4"
  role="button"
>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-0"
      type="Announcement"
    >
      <img
        src="test-file-stub"
      />
    </div>
    <span
      class="emotion-1"
    >
      New
    </span>
  </div>
  <p
    class="emotion-3"
  >
    Mock What's New article title
  </p>
</a>
`;
